### ISSUE TITLE
[selectors-4] Consider disallowing :has() outside the rightmost compound

### ISSUE THREAD

**Issue by anttijk** [w3c/csswg-drafts#6845]
Consider a rule like

`:has(.foo) div { ... }`

With this, style of every div in the document becomes dependent on the class lists of every element in the document. Specific cases can be optimized but in general case this means the engine has to do expensive full-document invalidation (or style recalc) on every document mutation (or pseudo class change) that may affect :has results. This is a potential performance footgun.

The invalidation scope can be drastically reduced by limiting :has() to rightmost compound only. This means that only ancestors/sibling chains needs to be considered for invalidation on :has()-affecting mutation. This is analogous to how only descendants/siblings need to be considered for regular selectors.

Sensible use cases for :has() are likely in the rightmost compound anyway so it seems better to not have this footgun around.

**Comment by Loirooriol** [comment]
Cc @byung-woo

**Comment by LeaVerou** [comment]
@anttijk Is WebKit implementing `:has()`??

**Comment by meyerweb** [comment]
> @anttijk Is WebKit implementing `:has()`??

Some prototyping in WebKit has happened and is ongoing: https://www.youtube.com/watch?v=bEcNxI0MYzk (progress as of the talk begins around 4:15)

**Comment by anttijk** [comment]
WebKit implementation is tracked in https://bugs.webkit.org/show_bug.cgi?id=227702.

**Comment by dbaron** [comment]
I think this almost makes `:has()` equivalent to the [old `:subject` proposal](https://www.w3.org/TR/2000/WD-css3-selectors-20001005/#subject-pseudo) (which was used to mark which piece of a selector was the subject), with the exception that you could still have more than one `:has()` on the rightmost piece-between-combinators.  (Or do you propose to limit that as well?  How much complexity and value does allowing more than one `:has()` yield?)

(I did suggest a few times that `:subject` might be preferable because it might be easier to implement; I found [one brief reference](https://lists.w3.org/Archives/Public/www-style/2019Jul/0002.html) but I think I'd written or said something longer at some point before that.)

**Comment by anttijk** [comment]
I don't think there is any particular need for other limitations. Multiple or nested :has() may be silly but it does not create new performance issuer or significantly increase implementation complexity.

**Comment by tabatkins** [comment]
I don't think it's unreasonable to have cases where you want it on something other than the last compound. For example, if you're using `:has()` to select a container, and then styling things *inside* the container.

This would happen quite naturally and often with Nesting, for instance.

**Comment by meyerweb** [comment]
> I don't think it's unreasonable to have cases where you want it on something other than the last compound. For example, if you're using `:has()` to select a container, and then styling things _inside_ the container.

I agree.  As an author, I don’t know that I would use this a lot, but I’m sure there would be instances of this coming up with tables and forms, to pick just two scenarios.  I have a gut feeling it would come up in styling Web Components as well.

**Comment by byung-woo** [comment]
I'm not sure that this understanding helps this discussion, but I think it depends on the authors intention.

If an author needs to use `:has(.foo) div { ... }`, it means that the intention of the author is styling every `div` in the document according to the class value `.foo` in any elements. If we place the limitation, then anyway the author need to select every `div` elements and style those elements with some JavaScript codes whenever the class value `.foo` is changed.

If the author's intention is to style `div` elements inside some elements with certain condition (e.g. `.bar`), then he can add additional condition for the `:has(.foo)` affecting element (e.g. `.bar:has(.foo) div {...}`) to make the invalidation scope smaller. 

Like this, I think that, there can be some options that authors can choose to handle some trade-offs.

**Comment by anttijk** [comment]
The `:has()` implantation in STP 137 does not have this restriction and I now think we probably can optimize all cases sufficiently without it.
